src/models/Integration.php

```diff
@@ -0,0 +1,379 @@
+    }
+    if ($team->getAdmin()) {
+      $redirect = 'admin';
+    } else {
+      $redirect = 'game';
+    }
+
+    $login_url = '/index.php?p='.$redirect;
+
+    return $login_url;
+  }
+
+  public static async function genFacebookOAuth(): Awaitable<bool> {
+    list($client, $url) = await self::genFacebookAuthURL("oauth");
+    $helper = $client->getRedirectLoginHelper();
+
+    $code = idx(Utils::getGET(), 'code');
+    $error = idx(Utils::getGET(), 'error');
+
+    if (!is_string($code)) {
+      $code = false;
+    }
+
+    if (!is_string($error)) {
+      $error = false;
+    }
+
+    $accessToken = '';
+
+    if ($code !== false) {
+      $graph_error = false;
+      try {
+        $accessToken = $helper->getAccessToken();
+      } catch (Facebook\Exceptions\FacebookResponseException $e) {
+        $graph_error = true;
+      } catch (Facebook\Exceptions\FacebookSDKException $e) {
+        $graph_error = true;
+      }
+
+      if ($graph_error === true) {
+        return false;
+      } else {
+        $response =
+          $client->get('/me?fields=id,third_party_id,email', $accessToken);
+        $profile = $response->getGraphUser();
+        $email = $profile['email'];
+        $id = $profile['third_party_id'];
+
+        $set_integrations = await self::genSetTeamIntegrations(
+          SessionUtils::sessionTeam(),
+          'facebook_oauth',
+          $email,
+          $id,
+        );
+        return $set_integrations;
+      }
+    } else if ($error !== false) {
+      return false;
+    }
+
+    header('Location: '.filter_var($url, FILTER_SANITIZE_URL));
+    exit;
+    return false;
+  }
+
+  public static async function genGoogleOAuth(): Awaitable<bool> {
+    list($client, $url) = await self::genGoogleAuthURL("oauth");
+
+    $code = idx(Utils::getGET(), 'code');
+    $error = idx(Utils::getGET(), 'error');
+
+    if (!is_string($code)) {
+      $code = false;
+    }
+
+    if (!is_string($error)) {
+      $error = false;
+    }
+
+    if ($code !== false) {
+      $client->authenticate($code);
+      $access_token = $client->getAccessToken();
+      $oauth_client = new Google_Service_Oauth2($client);
+      $profile = $oauth_client->userinfo->get();
+      $email = $profile->email;
+      $id = $profile->id;
+
+      $set_integrations = await self::genSetTeamIntegrations(
+        SessionUtils::sessionTeam(),
+        'google_oauth',
+        $email,
+        $id,
+      );
+      return $set_integrations;
+    } else if ($error !== false) {
+      return false;
+    }
+
+    header('Location: '.filter_var($url, FILTER_SANITIZE_URL));
+    exit;
+    return false;
+  }
+
+  public static async function genSetTeamIntegrations(
+    int $team_id,
+    string $type,
+    string $email,
+    string $id,
+  ): Awaitable<bool> {
+    $livesync_password_update =
+      await Team::genSetLiveSyncPassword($team_id, $type, $email, $id);
+
+    $oauth_token_update =
+      await Team::genSetOAuthToken($team_id, $type, $email);
```

Comment:
when independent:

```Hack
list($a, $b) = await \HH\Asio\va(foo::genA(), foo::genB());
```

Otherwise async just adds overhead, and can't get you anything.

Reply:
Absolutely!  We normally combine the awaitables and intend to do so in this PR.